File: src/Web/JavaScriptJS/builder.js
class TextReportBuilder {
  constructor() {
    this.parts = [];
  }

  reset() {
    this.parts = [];
  }

  addTitle(title) {
    this.parts.push(`# ${title}`);
  }

  addSection(heading, body) {
    this.parts.push(`## ${heading}`, body);
  }

  build() {
    return this.parts.join("\n");
  }
}

class HtmlReportBuilder {
  constructor() {
    this.parts = [];
  }

  reset() {
    this.parts = [];
  }

  addTitle(title) {
    this.parts.push(`<h1>${title}</h1>`);
  }

  addSection(heading, body) {
    this.parts.push(`<h2>${heading}</h2>`, `<p>${body}</p>`);
  }

  build() {
    return this.parts.join("");
  }
}

function buildAvailabilityReport(builder) {
  builder.reset();
  builder.addTitle("Service status");
  builder.addSection("Availability", "99.95%");
  return builder.build();
}

console.log(buildAvailabilityReport(new TextReportBuilder()));
console.log("---");
console.log(buildAvailabilityReport(new HtmlReportBuilder()));

File: src/Scripting/RubyRB/builder.rb
class TextReportBuilder
  def initialize
    @parts = []
  end

  def reset
    @parts.clear
  end

  def add_title(title)
    @parts << "# #{title}"
  end

  def add_section(heading, body)
    @parts << "## #{heading}"
    @parts << body
  end

  def build
    @parts.join("\n")
  end
end

class HtmlReportBuilder
  def initialize
    @parts = []
  end

  def reset
    @parts.clear
  end

  def add_title(title)
    @parts << "<h1>#{title}</h1>"
  end

  def add_section(heading, body)
    @parts << "<h2>#{heading}</h2>"
    @parts << "<p>#{body}</p>"
  end

  def build
    @parts.join
  end
end

def build_availability_report(builder)
  builder.reset
  builder.add_title('Service status')
  builder.add_section('Availability', '99.95%')
  builder.build
end

puts build_availability_report(TextReportBuilder.new)
puts '---'
puts build_availability_report(HtmlReportBuilder.new)
